README.md
# BBScan 1.4 #

A vulnerability scanner focus on scanning large number of targets in short time with a minimal set of rules.

**BBScan** 用于渗透测试前期，快速地对大量目标进行扫描，发现信息泄露等常见漏洞，找到可能的突破入口。 

它的特点是快速，规则配置简单。

## Change Log

* [2019-05-13]  BBScan 1.4 with scan strategy optimized.

## Install ##

Install required packages with pip

	pip install -r requirements.txt

## Usage ##

	usage: BBScan.py [options]
	
	* A tiny Batch weB+ vulnerability Scanner. *
	By LiJieJie (http://www.lijiejie.com)
	
	optional arguments:
	  -h, --help            show this help message and exit
	  --host [HOST [HOST2 HOST3 ...] [HOST [HOST2 HOST3 ...] ...]]
	                        Scan several hosts from command line
	  -f TargetFile         Load new line delimited targets from TargetFile
	  -d TargetDirectory    Load all *.txt files from TargetDirectory
	  --crawler CrawlDirectory
	                        Load all *.log crawl files from CrawlDirectory
	  --full                Process all sub directories.
	  -n, --no-crawl        No crawling, sub folders will not be processed.
	  -nn, --no-check404    No HTTP 404 existence check
	  --scripts-only        Scan with user scripts only
	  --no-scripts          Disable user scripts scan
	  -p PROCESS            Num of processes running concurrently, 30 by default
	  -t THREADS            Num of scan threads for each scan process, 3 by default
	  --network MASK        Scan all Target/MASK hosts,
	                        should be an int between 24 and 31
	  --timeout Timeout     Max scan minutes for each website, 10 by default
	  -nnn, --no-browser    Do not auto open web browser after scan finished
	  -md                   Save scan report as markdown format
	  -v                    show program's version number and exit


**1. Scan several hosts from command line** 

	python BBScan.py  --host www.a.com www.b.com

**2. Scan www.target.com and all the other IPs under www.target.com/28**

	python BBScan.py  --host www.target.com --network 28
	
**3. Load newline delimited targets from file and scan**
	
	python BBScan.py -f wandoujia.com.txt

**4. Load all targets from Directory(\*.txt file only) and scan**

	python BBScan.py -d targets/

**5. Load crawler logs from Directory(\*.log file only) and scan**

	python BBScan.py --crawler crawler_logs/

crawler log files should be formarted first:

			. GET http://www.iqiyi.com/ HTTP/1.1^^^200
			. POST http://www.pps.tv/login.php HTTP/1.1^^^user=admin&passwd=admin^^^200

rules/change_log.txt
# Chang Log
/readme               {status=200}      {type_no="html"}      {root_only}
/README               {status=200}      {type_no="html"}      {root_only}
/readme.md            {status=200}      {type_no="html"}      {root_only}
/readme.html          {status=200}      {type="html"}         {root_only}
/changelog.txt        {status=200}      {type="text/plain"}   {root_only}

lib/cmdline.py
#!/usr/bin/env python
#
#  Parse command line arguments
#


import argparse
import sys
import os


def parse_args():
    parser = argparse.ArgumentParser(prog='BBScan',
                                     formatter_class=argparse.RawTextHelpFormatter,
                                     description='* A tiny Batch weB vulnerability Scanner. *\n'
                                                 'By LiJieJie (http://www.lijiejie.com)',
                                     usage='BBScan.py [options]')

    parser.add_argument('--host', metavar='HOST [HOST2 HOST3 ...]', type=str, default='', nargs='*',
                        help='Scan several hosts from command line')

    parser.add_argument('-f', metavar='TargetFile', type=str, default='',
                        help='Load new line delimited targets from TargetFile')

    parser.add_argument('-d', metavar='TargetDirectory', type=str, default='',
                        help='Load all *.txt files from TargetDirectory')

    parser.add_argument('--crawler', metavar='CrawlDirectory', type=str, default='',
                        help='Load all *.log crawl files from CrawlDirectory')

    parser.add_argument('--full', dest='full_scan', default=False, action='store_true',
                        help='Process all sub directories.')

    parser.add_argument('-n', '--no-crawl', dest='no_crawl', default=False, action='store_true',
                        help='No crawling, sub folders will not be processed.')

    parser.add_argument('-nn', '--no-check404', dest='no_check404', default=False, action='store_true',
                        help='No HTTP 404 existence check')

    parser.add_argument('--scripts-only', dest='scripts_only', default=False, action='store_true',
                        help='Scan with user scripts only')

    parser.add_argument('--no-scripts', dest='no_scripts', default=False, action='store_true',
                        help='Disable user scripts scan')

    parser.add_argument('-p', metavar='PROCESS', type=int, default=30,
                        help='Num of processes running concurrently, 30 by default')

    parser.add_argument('-t', metavar='THREADS', type=int, default=3,
                        help='Num of scan threads for each scan process, 3 by default')

    parser.add_argument('--network', metavar='MASK', type=int, default=32,
                        help='Scan all Target/MASK hosts, \nshould be an int between 24 and 31')

    parser.add_argument('--timeout', metavar='Timeout', type=int, default=10,
                        help='Max scan minutes for each website, 10 by default')

    parser.add_argument('-nnn', '--no-browser', dest='no_browser', default=False, action='store_true',
                        help='Do not auto open web browser after scan finished')

    parser.add_argument('-md', default=False, action='store_true',
                        help='Save scan report as markdown format')

    parser.add_argument('-v', action='version', version='%(prog)s 1.4  By LiJieJie')

    if len(sys.argv) == 1:
        sys.argv.append('-h')

    args = parser.parse_args()
    check_args(args)
    return args


def check_args(args):
    if not (args.f or args.d or args.host or args.crawler):
        msg = 'Args missing! One of following args should be specified  \n' \
              '           -f TargetFile \n' \
              '           -d TargetDirectory \n' \
              '           --crawler TargetDirectory \n' \
              '           --host www.host1.com www.host2.com 8.8.8.8'
        print msg
        exit(-1)

    if args.f and not os.path.isfile(args.f):
        print '[ERROR] TargetFile not found: %s' % args.f
        exit(-1)

    if args.d and not os.path.isdir(args.d):
        print '[ERROR] TargetDirectory not found: %s' % args.f
        exit(-1)

    args.network = int(args.network)
    if not (24 <= args.network <= 32):
        print '[ERROR] Network should be an integer between 24 and 31'
        exit(-1)
